experimental/uefi/src/main.rs

```diff
@@ -51,14 +62,57 @@ fn _start(_handle: Handle, mut system_table: SystemTable<Boot>) -> Status {
         .reset(ResetType::Shutdown, status, None);
 }
 
-fn main(_handle: Handle, system_table: &mut SystemTable<Boot>) -> Status {
+fn main(handle: Handle, system_table: &mut SystemTable<Boot>) -> Status {
     use core::fmt::Write;
 
-    let status = writeln!(system_table.stdout(), "Hello World!");
+    writeln!(system_table.stdout(), "Hello World!").unwrap();
+
+    serial_echo(handle, system_table.boot_services()).unwrap();
+}
+
+fn echo_loop(serial: &mut uefi::proto::console::serial::Serial) -> Result<!, uefi::Error<()>> {
+    let mut buf: [u8; 1024] = [0; 1024];
+
+    loop {
+        // read() returns Ok if it managed to fill the whole buffer, or the error will contain
+        // the number of bytes read. The only error we're fine with is TIMEOUT, as we can simply
+        // retry that (and we'll keep getting TIMEOUTs when nobody is talking to us). In case of
+        // any other error, bail out.
+        let len = serial.read(&mut buf).map(|_| buf.len()).or_else(|err| {
+            if err.status() == Status::TIMEOUT {
+                Ok(*err.data())
+            } else {
+                Err(err.status())
+            }
+        })?;
+
+        // Write out what we read; if we get any errors, propagate them.
+        if len > 0 {
+            info!("Read data: {:?}", &buf[..len]);
+            serial.write(&buf[..len]).discard_errdata()?;
+        }
+    }
+}
+
+fn serial_echo(handle: Handle, bt: &BootServices) -> Result<!, uefi::Error<()>> {
+    // Expect (at least) two serial ports on the system; the first will be used
+    // for stdio, and we can use the second one for our echo example. If we don't
+    // seem to have a second serial port, err out with the (arbitrarily chosen)
+    // NO_MAPPING error.
+    let serial_handles = bt.find_handles::<Serial>()?;
+    let serial_handle = serial_handles.get(1).ok_or(Status::NO_MAPPING)?;
+    let serial = bt.open_protocol::<Serial>(
+        OpenProtocolParams {
+            handle: *serial_handle,
+            agent: handle,
+            controller: None,
+        },
+        OpenProtocolAttributes::Exclusive,
+    )?;
+    // Dereference the raw pointer we get to the serial interface.
+    let serial = unsafe { &mut *serial.interface.get() };
```

Comment:
Even if unsafe is necessary, I think it can probably be more confined, something like 
 
```
let serial = &mut unsafe { *serial.interface.get() };
```

Or does that not work?